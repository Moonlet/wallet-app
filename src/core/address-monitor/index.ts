import { Notifications } from '../messaging/notifications/notifications';
import { Blockchain } from '../blockchain/types';
import DeviceInfo from 'react-native-device-info';
import { IWalletsState, IAccountState } from '../../redux/wallets/state';
import { CONFIG } from '../../config';

const url = CONFIG.dataApiUrl + '/notifications/register';
const monitoredBlockchains = [Blockchain.ETHEREUM, Blockchain.ZILLIQA];

// const getDeviceToken = (): Promise<string> =>
//     Platform.OS === 'android' ? Notifications.getToken() : getApnsToken();

export const updateAddressMonitorTokens = (wallets: IWalletsState) => {
    const addresses = {};

    Object.keys(wallets || []).forEach(id => {
        wallets[id].accounts.forEach((account: IAccountState) => {
            if (monitoredBlockchains.indexOf(account.blockchain) !== -1) {
                if (!addresses[account.blockchain]) {
                    addresses[account.blockchain] = [];
                }

                addresses[account.blockchain].push(account.address);
            }
        });
    });

    const requestAddresses = Object.keys(addresses).map(blockchain => ({
        blockchain,
        addresses: addresses[blockchain]
    }));

    const deviceId = DeviceInfo.getUniqueId();
    Notifications.getToken().then(token => {
        const request = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                deviceId,
                tokenType: 'fcm',
                token,
                addresses: requestAddresses
            })
        };

        fetch(url, request);
    });
};
