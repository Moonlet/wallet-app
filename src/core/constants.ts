import { Blockchain } from './blockchain/types';

const BLOCKCHAIN_COINS = {
    [Blockchain.ETHEREUM]: 'ETH',
    [Blockchain.ZILLIQA]: 'ZIL',
    [Blockchain.COSMOS]: 'ATOM',
    [Blockchain.STELLAR]: 'XLM'
};

export { BLOCKCHAIN_COINS };
