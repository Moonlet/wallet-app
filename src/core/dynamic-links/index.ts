import dynamicLinks from '@react-native-firebase/dynamic-links';
import { URL, URLSearchParams } from 'react-native-url-polyfill';
import { captureException as SentryCaptureException } from '@sentry/react-native';
import { IReduxState } from '../../redux/state';
import { NavigationService } from '../../navigation/navigation-service';
import { getChainId } from '../../redux/preferences/selectors';
import { setSelectedBlockchain } from '../../redux/wallets/actions';
import { Blockchain } from '../blockchain/types';
import { getNrPendingTransactions, getSelectedAccount } from '../../redux/wallets/selectors';
import { ITokenState } from '../../redux/wallets/state';
import { store } from '../../redux/config';

export class DynamicLinksService {
    private onLinkListener = null;
    private state: IReduxState = null;

    private handlers = {
        '/:blockchain/:token/validator/:validatorAddress': params => {
            this.navigateValidatorAddress(params);
        },
        '/:blockchain/:token/stake/:validatorAddress': params => {
            this.stakeValidatorAddress(params);
        },
        '/:blockchain/stake': params => {
            this.blockchainStake(params);
        },
        '/stake': params => {
            // TODO: implement later on
            // console.log('stakeDefault', params);
        }
    };

    private navigateValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;
        const tokenSymbol = String(params?.token)?.toUpperCase();

        if (blockchain && validatorAddress && tokenSymbol) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            const chainId = getChainId(this.state, blockchain);

            const selectedAccount = getSelectedAccount(this.state);

            if (!selectedAccount) {
                SentryCaptureException(
                    new Error(
                        `Dynamic Links - no selected account, no wallet, blockchain: ${blockchain}, tokenSymbol: ${tokenSymbol}, validatorAddress: ${validatorAddress}`
                    )
                );
                return;
            }

            const token: ITokenState = selectedAccount.tokens[chainId][tokenSymbol];

            NavigationService.navigate('Validator', {
                blockchain,
                undefined, // validator
                accountIndex: selectedAccount.index,
                token,
                canPerformAction: !getNrPendingTransactions(this.state), // hasPendingTransactions
                options: {
                    validatorAddress,
                    tokenSymbol
                }
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - navigateValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private stakeValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;

        if (blockchain && validatorAddress) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'SelectStakeAccount',
                    key: 'select-stake-account',
                    params: {
                        validatorId: validatorAddress
                    }
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - stakeValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private blockchainStake(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;

        if (blockchain) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'StakeSelectValidator',
                    key: 'stake-now-select-validator'
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - blockchainStake, invalid params: ${params}`)
            );
        }
    }

    public async configure(state: IReduxState) {
        this.state = state;

        const dl = dynamicLinks();

        this.onLinkListener = dl.onLink(link => {
            this.handleDynamicLink(link);
        });

        // application is in a background state or has fully quit
        dl.getInitialLink().then(link => {
            this.handleDynamicLink(link);
        });
    }

    // app is in the foreground state
    public async handleDynamicLink(link: any) {
        if (link?.url) {
            this.processUrl(link.url, this.handlers, {
                protocol: ['https:'],
                host: [
                    'moonlet.io',
                    'moonlet.xyz',
                    'moonletwallet.com',
                    'moonlet.app',
                    'moonlet.link'
                ]
            });
        } else {
            // Invalid link url, link:${link}`
        }
    }

    public removeListeners() {
        typeof this.onLinkListener === 'function' && this.onLinkListener();
    }

    private pathPatternToRegexp(pathPattern: string) {
        let exp = '^';

        const pathParts = pathPattern.split('/').filter(Boolean);
        const capturingGroups = [];
        for (const part of pathParts) {
            exp += '\\/';
            if (part.indexOf(':') === 0) {
                exp += '([^\\/]*)';
                capturingGroups.push(part.replace(/^:/, ''));
            } else {
                exp += part;
            }
        }

        exp += '$';
        return {
            regexp: new RegExp(exp),
            capturingGroups
        };
    }

    private parsePath(path: string, pathPatternString: string) {
        const pathPattern = this.pathPatternToRegexp(pathPatternString);
        const match = path.match(pathPattern.regexp);

        if (match) {
            const params = {};
            for (let i = 0; i < pathPattern.capturingGroups.length; i++) {
                params[pathPattern.capturingGroups[i]] = match[i + 1];
            }
            return params;
        }

        return null;
    }

    private processPath(path: string, handlers: any, urlParams = {}) {
        const pathsPatterns = Object.keys(handlers);
        // console.log(path, handlers, urlParams);
        for (const pathPattern of pathsPatterns) {
            if (typeof handlers[pathPattern] === 'function') {
                const params = this.parsePath(path, pathPattern);
                if (params) {
                    return handlers[pathPattern]({
                        ...urlParams,
                        ...params
                    });
                }
            }
        }
    }

    private paramsToObject(entries: any) {
        const result = {};
        for (const [key, value] of entries) {
            // each 'entry' is a [key, value] tupple
            result[key] = value;
        }
        return result;
    }

    private processUrl(url: string, pathHandlers: any = this.handlers, whitelist = {}) {
        const u = new URL(url);

        for (const key of Object.keys(whitelist)) {
            if (u && typeof u[key] === 'string' && whitelist[key].indexOf(u[key]) < 0) {
                SentryCaptureException(
                    new Error(
                        JSON.stringify(
                            `whitelist not matched ${key}=${u[key]} not in whitelist: ${
                                typeof whitelist[key] === 'string'
                                    ? whitelist[key]
                                    : whitelist[key].join(', ')
                            }`
                        )
                    )
                );
                return null;
            }
        }

        let path = u?.pathname;

        if (path && String(path).startsWith('/links')) path = path.replace('/links', '');

        const searchParams = new URLSearchParams(url);

        return this.processPath(path, pathHandlers, this.paramsToObject(searchParams.entries()));
    }
}

export const DynamicLinks = new DynamicLinksService();
