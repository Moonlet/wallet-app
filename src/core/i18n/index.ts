export * from './translation/translate-component';
export * from './translation/translate';
