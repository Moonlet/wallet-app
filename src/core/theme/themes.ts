export enum Theme {
    dark = 'dark',
    light = 'light'
}
