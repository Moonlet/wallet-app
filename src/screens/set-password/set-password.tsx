import React, { useState } from 'react';
import { View, TouchableHighlight, TextInput } from 'react-native';
import { NavigationParams, NavigationScreenProp, NavigationState } from 'react-navigation';
import { Button } from '../../library/button/button';

import stylesProvider from './styles';
import { withTheme } from '../../core/theme/with-theme';
import { Text } from '../../library';
import { ITheme } from '../../core/theme/itheme';
import { Icon } from '../../components/icon';
import { translate } from '../../core/i18n';

export interface IProps {
    navigation: NavigationScreenProp<NavigationState, NavigationParams>;
    styles: ReturnType<typeof stylesProvider>;
    theme: ITheme;
}

const passwordStrengthRegex = /^(?=.*[A-Z])(?=.*[a-z])(?=.*\d)[A-Za-z\d!@#$%^&*(),.?":{}|<>]{10,}$/;

export const SetPasswordScreenComponent = (props: IProps) => {
    const [revealPassword, setRevealPassword] = useState({ password: false, confirm: false });
    const [password, setPassword] = useState('');
    const [confirmPassword, setConfirmPassword] = useState('');
    const [errorText, setErrorText] = useState('');

    return (
        <View style={props.styles.container}>
            <View style={props.styles.topContainer}>
                <Text darker style={{ textAlign: 'center', marginTop: 60 }}>
                    {translate('SetPassword.body')}
                </Text>
                {errorText !== '' && (
                    <Text style={{ textAlign: 'center', marginTop: 20 }}>{errorText}</Text>
                )}
                <View style={props.styles.inputBox}>
                    <TextInput
                        testID="input-password"
                        style={props.styles.input}
                        placeholderTextColor={props.theme.colors.textSecondary}
                        placeholder={translate('SetPassword.password')}
                        autoCapitalize={'none'}
                        autoCorrect={false}
                        selectionColor={props.theme.colors.accent}
                        secureTextEntry={!revealPassword.password}
                        value={password}
                        onChangeText={text => {
                            setPassword(text);
                        }}
                    />
                    <TouchableHighlight
                        testID="reveal-password"
                        onPress={() => {
                            setRevealPassword({
                                ...revealPassword,
                                ...{ password: !revealPassword.password }
                            });
                        }}
                    >
                        <Icon
                            name={revealPassword.password ? 'view-1' : 'view-off'}
                            size={18}
                            style={props.styles.icon}
                        />
                    </TouchableHighlight>
                </View>

                <View style={props.styles.inputBox}>
                    <TextInput
                        testID="input-confirm-password"
                        style={props.styles.input}
                        placeholderTextColor={props.theme.colors.textSecondary}
                        placeholder={translate('SetPassword.confirmPassword')}
                        autoCapitalize={'none'}
                        autoCorrect={false}
                        selectionColor={props.theme.colors.accent}
                        secureTextEntry={!revealPassword.confirm}
                        value={confirmPassword}
                        onChangeText={text => {
                            setConfirmPassword(text);
                        }}
                    />
                    <TouchableHighlight
                        testID="reveal-confirm-password"
                        onPress={() => {
                            setRevealPassword({
                                ...revealPassword,
                                ...{ confirm: !revealPassword.confirm }
                            });
                        }}
                    >
                        <Icon
                            name={revealPassword.confirm ? 'view-1' : 'view-off'}
                            size={18}
                            style={props.styles.icon}
                        />
                    </TouchableHighlight>
                </View>
            </View>
            <View style={props.styles.bottomContainer}>
                <Button
                    testID="button-understand"
                    style={props.styles.bottomButton}
                    primary
                    onPress={() => {
                        if (password !== confirmPassword) {
                            setErrorText(translate('SetPassword.errors.passwordsDontMatch'));
                            return;
                        }

                        if (!passwordStrengthRegex.test(password)) {
                            setErrorText(translate('Password.invalidPassword'));
                            return;
                        }

                        // props.navigation.navigate('');
                    }}
                >
                    {translate('App.labels.secure')}
                </Button>
            </View>
        </View>
    );
};

const navigationOptions = () => ({
    title: translate('Wallets.secureWallet')
});

export const SetPasswordScreen = withTheme(stylesProvider)(SetPasswordScreenComponent);

SetPasswordScreen.navigationOptions = navigationOptions;
