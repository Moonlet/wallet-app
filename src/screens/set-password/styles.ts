import { StyleSheet } from 'react-native';
import { ITheme } from '../../core/theme/itheme';

export default (theme: ITheme) =>
    StyleSheet.create({
        container: {
            paddingHorizontal: 16,
            alignItems: 'center',
            justifyContent: 'space-between',
            backgroundColor: theme.colors.appBackground,
            ...StyleSheet.absoluteFillObject
        },

        topContainer: {
            flex: 1,
            justifyContent: 'flex-start',
            alignItems: 'center',
            alignSelf: 'stretch'
        },

        bottomContainer: {
            flex: 0,
            justifyContent: 'center',
            alignSelf: 'stretch',
            alignItems: 'center',
            marginBottom: 60
        },

        bottomButton: {
            width: '80%',
            marginTop: 40
        },

        icon: {
            color: theme.colors.accent,
            padding: 4
        },

        input: {
            flex: 1,
            color: theme.colors.text
        },

        inputBox: {
            height: 40,
            borderRadius: 6,
            borderColor: 'gray',
            alignSelf: 'stretch',
            backgroundColor: theme.colors.inputBackground,
            paddingHorizontal: 12,
            marginTop: 40,
            flexDirection: 'row',
            justifyContent: 'space-between',
            alignItems: 'center'
        }
    });
