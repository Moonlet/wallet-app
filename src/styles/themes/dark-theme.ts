import { ITheme } from '../../core/theme/itheme';
import { COLORS } from '../colors';

export const darkTheme: ITheme = {
    dark: true,
    shadowGradient: ['#00000000', '#000000CC'],

    fontSize: {
        small: 12,
        regular: 16,
        large: 34
    },

    colors: {
        primary: COLORS.GRAY,
        accent: COLORS.AQUA,
        text: COLORS.WHITE,
        textDarker: COLORS.LIGHT_GRAY,
        positive: COLORS.HOT_GREEN,
        negative: COLORS.HOT_RED,

        cardBackground: COLORS.DARK_GRAY,
        appBackground: COLORS.COD_GRAY,
        settingsDivider: COLORS.BLACK
    }
};
