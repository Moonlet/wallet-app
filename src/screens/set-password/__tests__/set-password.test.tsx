import React from 'react';
import { SetPasswordScreenComponent, IProps } from '../set-password';
import { darkTheme } from '../../../styles/themes/dark-theme';
import styleProvider from '../styles';

import { shallow } from 'enzyme';
import { loadTranslations } from '../../../core/i18n';

const props: IProps = {
    // @ts-ignore
    navigation: {
        navigate: jest.fn()
    },
    styles: styleProvider(darkTheme),
    theme: darkTheme
};

describe('set password screen component', () => {
    beforeAll(async () => {
        await loadTranslations('en');
    });

    it('renders correctly', () => {
        const wrapper = shallow(<SetPasswordScreenComponent {...props} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('sets password correctly', () => {
        const wrapper = shallow(<SetPasswordScreenComponent {...props} />);

        wrapper.find('[testID="input-password"]').simulate('changeText', 'pass1');
        wrapper.find('[testID="input-confirm-password"]').simulate('changeText', 'pass1');
        wrapper.find('[testID="reveal-password"]').simulate('press');
        wrapper.find('[testID="reveal-confirm-password"]').simulate('press');

        expect(wrapper.debug()).toMatchSnapshot();
    });
});
