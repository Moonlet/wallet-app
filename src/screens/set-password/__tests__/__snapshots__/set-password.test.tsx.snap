// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set password screen component renders correctly 1`] = `
"<View style={{...}}>
  <View style={{...}}>
    <withTheme(TextComponent) darker={true} style={{...}}>
      10 or more characters long, at least 1 number, 1 lowercase character and UPPERCASE character.
    </withTheme(TextComponent)>
    <View style={{...}}>
      <TextInput testID=\\"input-password\\" style={{...}} placeholderTextColor=\\"#9a9a9a\\" placeholder=\\"password\\" autoCapitalize=\\"none\\" autoCorrect={false} selectionColor=\\"#00DAFF\\" secureTextEntry={true} value=\\"\\" onChangeText={[Function: onChangeText]} allowFontScaling={true} rejectResponderTermination={true} underlineColorAndroid=\\"transparent\\" />
      <TouchableHighlight testID=\\"reveal-password\\" onPress={[Function: onPress]} activeOpacity={0.85} delayPressOut={100} underlayColor=\\"black\\">
        <Icon name=\\"view-off\\" size={18} style={{...}} allowFontScaling={false} />
      </TouchableHighlight>
    </View>
    <View style={{...}}>
      <TextInput testID=\\"input-confirm-password\\" style={{...}} placeholderTextColor=\\"#9a9a9a\\" placeholder=\\"confirm password\\" autoCapitalize=\\"none\\" autoCorrect={false} selectionColor=\\"#00DAFF\\" secureTextEntry={true} value=\\"\\" onChangeText={[Function: onChangeText]} allowFontScaling={true} rejectResponderTermination={true} underlineColorAndroid=\\"transparent\\" />
      <TouchableHighlight testID=\\"reveal-confirm-password\\" onPress={[Function: onPress]} activeOpacity={0.85} delayPressOut={100} underlayColor=\\"black\\">
        <Icon name=\\"view-off\\" size={18} style={{...}} allowFontScaling={false} />
      </TouchableHighlight>
    </View>
  </View>
  <View style={{...}}>
    <withTheme(ButtonComponent) testID=\\"button-understand\\" style={{...}} primary={true} onPress={[Function: onPress]}>
      Secure
    </withTheme(ButtonComponent)>
  </View>
</View>"
`;

exports[`set password screen component sets password correctly 1`] = `
"<View style={{...}}>
  <View style={{...}}>
    <withTheme(TextComponent) darker={true} style={{...}}>
      10 or more characters long, at least 1 number, 1 lowercase character and UPPERCASE character.
    </withTheme(TextComponent)>
    <View style={{...}}>
      <TextInput testID=\\"input-password\\" style={{...}} placeholderTextColor=\\"#9a9a9a\\" placeholder=\\"password\\" autoCapitalize=\\"none\\" autoCorrect={false} selectionColor=\\"#00DAFF\\" secureTextEntry={false} value=\\"pass1\\" onChangeText={[Function: onChangeText]} allowFontScaling={true} rejectResponderTermination={true} underlineColorAndroid=\\"transparent\\" />
      <TouchableHighlight testID=\\"reveal-password\\" onPress={[Function: onPress]} activeOpacity={0.85} delayPressOut={100} underlayColor=\\"black\\">
        <Icon name=\\"view-1\\" size={18} style={{...}} allowFontScaling={false} />
      </TouchableHighlight>
    </View>
    <View style={{...}}>
      <TextInput testID=\\"input-confirm-password\\" style={{...}} placeholderTextColor=\\"#9a9a9a\\" placeholder=\\"confirm password\\" autoCapitalize=\\"none\\" autoCorrect={false} selectionColor=\\"#00DAFF\\" secureTextEntry={false} value=\\"pass1\\" onChangeText={[Function: onChangeText]} allowFontScaling={true} rejectResponderTermination={true} underlineColorAndroid=\\"transparent\\" />
      <TouchableHighlight testID=\\"reveal-confirm-password\\" onPress={[Function: onPress]} activeOpacity={0.85} delayPressOut={100} underlayColor=\\"black\\">
        <Icon name=\\"view-1\\" size={18} style={{...}} allowFontScaling={false} />
      </TouchableHighlight>
    </View>
  </View>
  <View style={{...}}>
    <withTheme(ButtonComponent) testID=\\"button-understand\\" style={{...}} primary={true} onPress={[Function: onPress]}>
      Secure
    </withTheme(ButtonComponent)>
  </View>
</View>"
`;
